File: be-api/src/main/java/cd/beapi/service/MedicineService.java
package cd.beapi.service;

import cd.beapi.dto.request.CreateMedicineRequest;
import cd.beapi.dto.request.SearchMedicineRequest;
import cd.beapi.dto.request.UpdateMedicineRequest;
import cd.beapi.dto.response.MedicineResponse;
import cd.beapi.dto.response.PageData;
import org.springframework.core.io.Resource;

public interface MedicineService {
    MedicineResponse findById(Long id);
    PageData<MedicineResponse> search(SearchMedicineRequest searchMedicineRequest);
    MedicineResponse save(CreateMedicineRequest createMedicineRequest);
    MedicineResponse update(Long id, UpdateMedicineRequest updateMedicineRequest);
    void delete(Long id);
    Resource exportExcel();
}

File: fe-admin/src/app/core/service/medicine.service.ts
import {inject, Injectable} from '@angular/core';
import {ENV} from "../../environment";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {SuccessResponse} from "../model/response/success-response";
import {PageData} from "../model/response/page-data";
import {SearchMedicineRequest} from "../model/request/search-medicine-request";
import {MedicineResponse} from "../model/response/medicine-response";
import {CreateMedicineRequest} from "../model/request/create-medicine-request";
import {UpdateMedicineRequest} from "../model/request/update-medicine-request";

@Injectable({
    providedIn: 'root',
})
export class MedicineService {
    private readonly url = ENV.API_BASE_URL + "clinic/medicines";
    private readonly http = inject(HttpClient);

    search(searchMedicineRequest: SearchMedicineRequest): Observable<SuccessResponse<PageData<MedicineResponse>>> {
        return this.http.post<SuccessResponse<PageData<MedicineResponse>>>(`${this.url}/search`, searchMedicineRequest);
    }

    findById(id: string): Observable<SuccessResponse<MedicineResponse>> {
        return this.http.get<SuccessResponse<MedicineResponse>>(`${this.url}/${id}`);
    }

    create(createMedicineRequest: CreateMedicineRequest): Observable<SuccessResponse<MedicineResponse>> {
        return this.http.post<SuccessResponse<MedicineResponse>>(this.url, createMedicineRequest);
    }

    update(id: string, updateMedicineRequest: UpdateMedicineRequest): Observable<SuccessResponse<MedicineResponse>> {
        return this.http.put<SuccessResponse<MedicineResponse>>(`${this.url}/${id}`, updateMedicineRequest);
    }

    delete(id: string): Observable<SuccessResponse<void>> {
        return this.http.delete<SuccessResponse<void>>(`${this.url}/${id}`);
    }

    exportExcel(): Observable<Blob> {
        return this.http.get(`${this.url}/export`, {responseType: 'blob'});
    }
}

